tokio/src/runtime/thread_pool/worker.rs

```diff
@@ -538,42 +535,25 @@ impl Core {
         true
     }
 
-    /// Runs maintenance work such as free pending tasks and check the pool's
-    /// state.
+    /// Runs maintenance work such as checking the pool's state.
     fn maintenance(&mut self, worker: &Worker) {
-        self.drain_pending_drop(worker);
-
         if !self.is_shutdown {
             // Check if the scheduler has been shutdown
             self.is_shutdown = worker.inject().is_closed();
         }
     }
 
-    // Signals all tasks to shut down, and waits for them to complete. Must run
-    // before we enter the single-threaded phase of shutdown processing.
+    /// Signals all tasks to shut down, and waits for them to complete. Must run
+    /// before we enter the single-threaded phase of shutdown processing.
     fn pre_shutdown(&mut self, worker: &Worker) {
         // Signal to all tasks to shut down.
-        for header in self.tasks.iter() {
+        while let Some(header) = worker.shared.owned.pop_back() {
             header.shutdown();
         }
```

Comment:
Here all the threads will clean up tasks in parallel. I'm not sure what is best here.

Reply:
It is the shutdown process and not performance-sensitive. This seems fine to me.